wallet/udb/txmined.go

```diff
@@ -321,6 +321,99 @@ func (s *Store) UpdateProcessedTxsBlockMarker(dbtx walletdb.ReadWriteTx, hash *c
 	return nil
 }
 
+// Birthday returns the wallet's birthday time and a blockhash before that time
+// if set. If not set nil values are returned.
+func (s *Store) Birthday(dbtx walletdb.ReadTx) (*chainhash.Hash, *time.Time) {
+	const timeBits = 8
+	ns := dbtx.ReadBucket(wtxmgrBucketKey)
+	v := ns.Get(rootBirthdayHashAndTime)
+	if len(v) != chainhash.HashSize+timeBits {
+		return nil, nil
+	}
+	var h chainhash.Hash
+	copy(h[:], v[:chainhash.HashSize])
+	t := time.Unix(int64(byteOrder.Uint64(v[chainhash.HashSize:])), 0)
+	return &h, &t
+}
+
+// SetBirthday sets a birthday hash and time from a birthday time. The hash will
+// be from the block before the birthday time.
+//
+//	[0:32]  Block header hash (32 bytes)
+//	[32:40] Birthday time (8 bytes)
+func (s *Store) SetBirthday(dbtx walletdb.ReadWriteTx, birthday *time.Time) (*chainhash.Hash, error) {
```

Comment:
This looks like a mess, so if there is anything simpler to do, please make a suggestion. It goes backwards using an approximation based on the amount of time to the birthday compared to the lifetime of the chain. Then it goes forward one at a time until it goes past the time we want and uses the header before that.

Reply:
I really think the birthday should internally be just "birthblock" (height+hash), and then that can be specified and recorded precisely. 

You can then have multiple ways of specifying the birthblock when creating the wallet: by height or hash directly, by timestamp (like you do here), by "decred day" (like [aezeed does](https://github.com/decred/dcrlnd/tree/master/aezeed#plaintext-aezeed-encoding)) or any other method.

As you notice, you'll need two sets of states: the user-specified birthday, set  when invoking with `--create` (e.g. `--birthtimestamp`, `--birthheight`, etc) and then a separate state that records the _actual_ determined birthday hash and height (because you can't determine the hash before sync and sync may take an arbitrary amount of time to complete).

Instead of doing this process here, this can probably (haven't thought this through completely) be done inside the main db tx of `wallet.ChainSwitch()`: if the birthblock hash is not set, then go through the logic of determining the birthday within the set of the batch of headers being processed. This has the advantage of working both for SPV and RPC sync modes, avoids going through this process multiple times and meshes perfectly with the initial sync process, such that by the end of the initial sync, the birthblock will be set and ready to be used as the initial rescan point.

Reply:
Thank you! Doing what you said in `wallet.ChainSwitch()` I think. Please look again, the changes also https://github.com/decred/dcrwallet/compare/e52f2f26d5e210dde5d0b1c01aa1b8a46f4d4099..80145bce22dd5604c93e2c6e25cd5c72a0063885

Now allowing the consumer to set the intent to set a birthday, so two separate values in the db now.

If this looks ok, I guess we still want to use the birthday for rescans with default values, so I will add that if I can.